src/amo/components/PermissionsCard/index.js

```diff
@@ -0,0 +1,280 @@
+      // first, then individual host permissions.
+
+      if (allUrls) {
+        permissionsToDisplay.push(
+          <Permission
+            className="hostPermission"
+            description={permissionStrings.allUrls}
+            key="allUrls"
+          />
+        );
+      } else {
+        // Formats a list of host permissions.  If we have 4 or fewer, display
+        // them all, otherwise display the first 3 followed by an item that
+        // says "...plus N others".
+        const formatHostPermissions = (list, itemKey, moreKey) => {
+          function formatItems(items) {
+            permissionsToDisplay.push(...items.map((item) => {
+              return (
+                <Permission
+                  className="hostPermission"
+                  description={permissionStrings[itemKey](item)}
+                  key={item}
+                />
+              );
+            }));
+          }
+          if (list.length < 5) {
+            formatItems(list);
+          } else {
+            formatItems(list.slice(0, 3));
+
+            const remaining = list.length - 3;
+            permissionsToDisplay.push(
+              <Permission
+                className="hostPermission"
+                description={permissionStrings[moreKey](remaining)}
+                key={moreKey}
+              />
+            );
+          }
+        };
+
+        formatHostPermissions(wildcards, 'wildcard', 'tooManyWildcards');
+        formatHostPermissions(sites, 'oneSite', 'tooManySites');
+      }
+
+      // Next, show the native messaging permission if it is present.
+      const nativeMessagingPermission = 'nativeMessaging';
+      if (permissions.permissions.includes(nativeMessagingPermission)) {
+        permissionsToDisplay.push(
+          <Permission
+            className={nativeMessagingPermission}
+            description={permissionStrings[nativeMessagingPermission]}
+            key={nativeMessagingPermission}
+          />
+        );
+      }
+
+      // Finally, show remaining permissions, sorted alphabetically by the
+      // permission string to match Firefox.
+      const permissionsCopy = permissions.permissions.slice(0);
+      // eslint-disable-next-line no-restricted-syntax
+      for (const permission of permissionsCopy.sort()) {
+        // nativeMessaging is handled above.
+        if (permission === 'nativeMessaging') {
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+        // Only output a permission if we have a string defined for it.
+        if (permissionStrings[permission]) {
+          permissionsToDisplay.push(
+            <Permission
+              className={permission}
+              description={permissionStrings[permission]}
+              key={permission}
+            />
+          );
+        }
+      }
+      return permissionsToDisplay;
+    }
+
+    // Get the permission strings from the appropriate file object.
```

Comment:
I originally had this as a separate function, hoping that I could unit test it (as discussed above), but then I just inlined it. It's definitely missing test coverage currently and the best way to address that might be to extract it out into a helper as I discussed above with `classifyPermission`.